pkg/kds/mux/zone_watch.go

```diff
@@ -0,0 +1,97 @@
+package mux
+
+import (
+	"context"
+	"time"
+
+	"github.com/kumahq/kuma/pkg/core"
+	"github.com/kumahq/kuma/pkg/core/resources/apis/system"
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	"github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/core/resources/store"
+	"github.com/kumahq/kuma/pkg/events"
+	"github.com/kumahq/kuma/pkg/kds/service"
+	"github.com/kumahq/kuma/pkg/multitenant"
+)
+
+type zoneTenant struct {
+	zone     string
+	tenantID string
+}
+
+type ZoneWatch struct {
+	Bus   events.EventBus
+	RM    manager.ReadOnlyResourceManager
+	Poll  time.Ticker
+	zones map[zoneTenant]time.Time
+}
+
+func NewZoneWatch(
+	bus events.EventBus,
+	rm manager.ReadOnlyResourceManager,
+	poll time.Ticker,
+) *ZoneWatch {
+	return &ZoneWatch{
+		Bus:   bus,
+		RM:    rm,
+		Poll:  poll,
+		zones: map[zoneTenant]time.Time{},
+	}
+}
+
+func (zw *ZoneWatch) Start(stop <-chan struct{}) error {
+	defer zw.Poll.Stop()
+
+	core.Log.Info("STARTED")
+	connectionWatch := zw.Bus.Subscribe(func(e events.Event) bool {
+		_, ok := e.(service.ZoneOpenedStream)
+		return ok
+	})
+	defer connectionWatch.Close()
+
+	for {
+		select {
+		case <-zw.Poll.C:
+			core.Log.Info("new poll!")
+			for zone, firstSeen := range zw.zones {
+				ctx := multitenant.WithTenant(context.TODO(), zone.tenantID)
+				zoneInsight := system.NewZoneInsightResource()
+				if err := zw.RM.Get(ctx, zoneInsight, store.GetByKey(zone.zone, model.NoMesh)); err != nil {
+					//
+				}
+				lastHealthCheck := zoneInsight.Spec.GetHealthCheck().GetTime().AsTime()
+				if lastHealthCheck.After(firstSeen) && time.Since(lastHealthCheck) > 20*time.Second {
```

Comment:
race condition: zone connects but immediately disconnects before it can send a health check